### PR TITLE
[python] Sanitize the weights in the computations of the perplexity

### LINKED ISSUE #336: Nan in the computation of perpexlity
Due to numerical approximations, components with very small weighs yield errors in PMC sampling: `RuntimeWarning: divide by zero encountered in log` due to the line
https://github.com/eos/eos/blob/edce93909e4632fb646876063002ec54a5de17e0/python/eos/analysis.py#L407
When the normalized weight is zero, the contribution to the entropy should simply be zero. This is supposed to be accounted for by the previous lines, where non-positive weights are set to a small value `eps`:https://github.com/eos/eos/blob/edce93909e4632fb646876063002ec54a5de17e0/python/eos/analysis.py#L404-L405
I think that this step should be performed **after** the normalization and not **before**.
Idem here: https://github.com/eos/eos/blob/edce93909e4632fb646876063002ec54a5de17e0/python/eos/analysis.py#L432

@dvandyk Do you agree?

### PR DESCRIPTION
**Pull request by mreboud**
I tested this locally, should we think about a test case for this?
Closes #336.

### CODE REVIEW

File: python/eos/analysis.py [hunk 1]
```diff
@@ -400,19 +400,29 @@ def sample_pmc(self, log_proposal, step_N=1000, steps=10, final_N=5000, rng=np.r
             generating_components.append(origins)
             samples = sampler.samples[:]
             last_weights = np.copy(sampler.weights[-1][:, 0])
-            for i, w in enumerate(last_weights):
-                if w <= 0 or np.isnan(w):
-                    last_weights[i] = eps
-            normalized_last_weights = last_weights / np.sum(last_weights)
+            normalized_last_weights = last_weights / np.sum(
```
Comment (dvandyk) [inline python/eos/analysis.py:403]: This fails if none of the weights is positive, which is a pathological corner case. Should we test for it and skip the computation of the perplexity in that case?

File: python/eos/analysis.py [hunk 2]
```diff
@@ -425,10 +435,15 @@ def sample_pmc(self, log_proposal, step_N=1000, steps=10, final_N=5000, rng=np.r
...
-        normalized_weights = adjusted_weights / np.sum(adjusted_weights)
+        normalized_weights = adjusted_weights / np.sum(
+            adjusted_weights,
+            where = np.logical_or(adjusted_weights > 0, np.isfinite(adjusted_weights))
+        ) # sum positive finite weights only
+        normalized_weights = np.where(
+                np.logical_or(normalized_weights <= 0, np.isnan(normalized_weights)),
+                eps,
+                normalized_weights
+        ) # replace negative and nan weights by eps
```
Comment (dvandyk) [inline python/eos/analysis.py:446]: Seeing the same code thrice, should we make it a separaste function? Also, is there any benefit to using [pypmc's function ```pypmc.tools.perp```](https://pypmc.github.io/api.html#pypmc.tools.convergence.perp)?

A quick look at that source code seems better for the zero weights, but has not a workable solution for NaN and negative weights.

**Review by dvandyk — COMMENTED** [review]
Could be merged as is, might be improved.

**Comment by dvandyk** [comment]
Also, there is a typo in the commit message: ```perpexlity``` -> ```perplexity```

File: python/eos/analysis.py (hunk 2 above)
Comment (mreboud) [reply python/eos/analysis.py:446]: Indeed the pypmc's `perp` doesn't treat nan weights, that are even propagated due to the normalization. We could write our own `perp` function.

File: python/eos/analysis.py (hunk 1 above)
Comment (mreboud) [reply python/eos/analysis.py]: Yes, I set perplexity to zero in this case

**Comment by dvandyk** [comment]
Please go ahead with merging if the tests pass.
